File: domain/src/main/java/mixter/domain/core/subscription/handlers/NotifyFollowerOfFolloweeMessage.java
package mixter.domain.core.subscription.handlers;

import mixter.domain.EventPublisher;
import mixter.domain.core.message.MessageId;
import mixter.domain.core.message.events.MessagePublished;
import mixter.domain.core.message.events.MessageRepublished;
import mixter.domain.core.message.events.ReplyMessagePublished;
import mixter.domain.core.subscription.FollowerRepository;
import mixter.domain.core.subscription.Subscription;
import mixter.domain.core.subscription.SubscriptionId;
import mixter.domain.core.subscription.SubscriptionRepository;
import mixter.domain.identity.UserId;

import java.util.Set;

public class NotifyFollowerOfFolloweeMessage {

    private final FollowerRepository followerRepository;
    private final SubscriptionRepository subscriptionRepository;
    private final EventPublisher eventPublisher;

    public NotifyFollowerOfFolloweeMessage(FollowerRepository followerRepository, SubscriptionRepository subscriptionRepository, EventPublisher eventPublisher) {

        this.followerRepository = followerRepository;
        this.subscriptionRepository = subscriptionRepository;
        this.eventPublisher = eventPublisher;
    }

    public void apply(MessagePublished messagePublished) {
        notifyFollowers(messagePublished.getAuthorId(), messagePublished.getMessageId());
    }

    private void notifyFollowers(UserId followeee, MessageId messageId) {
        Set<UserId> followers = followerRepository.getFollowers(followeee);
        for (UserId follower : followers) {
            Subscription subscription = subscriptionRepository.getById(new SubscriptionId(follower, followeee));
            subscription.notifyFollower(messageId, eventPublisher);
        }
    }

    public void apply(MessageRepublished event) {
        notifyFollowers(event.getAuthorId(), event.getMessageId());
    }

    public void apply(ReplyMessagePublished messageReplied) {
        notifyFollowers(messageReplied.getAuthorId(), messageReplied.getMessageId());
    }
}

File: domain/src/main/scala/mixter/domain/subscription/handlers/NotifyFollowerOfFolloweeMessage.scala
package mixter.domain.subscription.handlers

import mixter.domain.EventPublisher
import mixter.domain.message.event.{MessageQuacked, MessageRequacked}
import mixter.domain.subscription.{FollowerRepository, SubscriptionId, SubscriptionRepository}

class NotifyFollowerOfFolloweeMessage(followerRepository: FollowerRepository,
                                      subscriptionRepository: SubscriptionRepository
                                     )(implicit val ep: EventPublisher) {

  def apply(messageQuacked: MessageQuacked): Unit = {
    for {
      follower <- followerRepository.getFollowers(messageQuacked.author)
      subscription <- subscriptionRepository.getById(SubscriptionId(follower, messageQuacked.author))
    } subscription.notifyFollower(messageQuacked.id)
  }
  def apply(messageRequacked: MessageRequacked): Unit = {
    for {
      follower <- followerRepository.getFollowers(messageRequacked.requacker)
      subscription <- subscriptionRepository.getById(SubscriptionId(follower, messageRequacked.requacker))
    } subscription.notifyFollower(messageRequacked.id)
  }

}
